app/gui2/src/util/containers.ts

```diff
@@ -0,0 +1,246 @@
+import { assertDefined, assertEqual } from '@/util/assert'
+import { expect, test } from 'vitest'
+
+/**
+ * Map that supports Object-based keys.
+ *
+ * Internally keys are converted to strings using the provided {@link keyMapper} function and then compared.
+ *
+ * @template Key The type of the keys.
+ * @template Value The type of the values.
+ */
+export class ObjectKeyedMap<Key extends Object, Value> {
+  /** The inner map that stores the values. */
+  private readonly map = new Map<string, [Key, Value]>()
+
+  /** Construct a new map, optionally setting a custom key mapper.
+   *
+   * @param keyMapper The function that maps the keys to strings. By default, strings are generated using {@link JSON.stringify}.
+   *
+   */
+  constructor(private readonly keyMapper: (key: Key) => string = JSON.stringify) {}
+
+  /** Set the value for the given key. */
+  set(key: Key, value: Value): this {
+    const innerKey = this.keyMapper(key)
+    this.map.set(innerKey, [key, value])
+    return this
+  }
+
+  /** Get the value for the given key, or `undefined` if it does not exist. */
+  get(key: Key): Value | undefined {
+    const innerKey = this.keyMapper(key)
+    const entry = this.map.get(innerKey)
+    return entry ? entry[1] : undefined
+  }
+
+  /** Check if the map contains a value for the given key. */
+  has(key: Key): boolean {
+    const innerKey = this.keyMapper(key)
+    return this.map.has(innerKey)
+  }
+
+  /** Remove the value for the given key. */
+  delete(key: Key): boolean {
+    const innerKey = this.keyMapper(key)
+    return this.map.delete(innerKey)
+  }
+
+  /** Remove all values from the map. */
+  clear(): void {
+    this.map.clear()
+  }
+
+  /** Get the number of values in the map. */
+  get size(): number {
+    return this.map.size
+  }
+
+  /** Iterate over the values in the map. */
+  *[Symbol.iterator](): IterableIterator<[Key, Value]> {
+    for (const [_innerKey, [key, value]] of this.map.entries()) {
+      yield [key, value]
+    }
+  }
+}
+
+/**
+ * Set that supports Object-based keys.
+ *
+ * Internally keys are converted to strings using the provided {@link keyMapper} function and then compared.
+ *
+ * @template Key The type of the keys.
+ */
+export class ObjectKeyedSet<Key extends Object> {
```

Comment:
Sets are usually phrased as "set of values", not "set of keys". Also, this class doesn't allow you to iterate over the actual values stored in the set, but only over the mapped keys. That seems strange to me, especially since the map actually gives you the original keys back.

Reply:
Fixed.